crates/fingerprint/src/benchmark.rs

```diff
@@ -0,0 +1,117 @@
+    /// different sizes.
+    pub size: u64,
+}
+
+impl Benchmark {
+    pub fn fingerprint<P: AsRef<Path>>(path: P) -> Result<Self> {
+        let path = path.as_ref().canonicalize()?;
+
+        // Calculate the hash for the benchmark file.
+        let size = File::open(&path)
+            .expect("should be able to open the benchmark file")
+            .metadata()
+            .expect("should be able to collect the benchmark file size")
+            .len();
+
+        Ok(Self {
+            name: simplify_benchmark_name(&path)?,
+            path: simplify_benchmark_path(&path),
+            hash: hash::file(&path),
+            size,
+        })
+    }
+}
+
+/// Simplify the benchmark name if possible; e.g.:
+/// - `.../<name>/benchmark.wasm` -> <name>
+/// - `.../<name>.wasm -> <name>
+fn simplify_benchmark_name<P: AsRef<Path>>(path: P) -> Result<String> {
+    let path = path.as_ref();
+    let stem = path
+        .file_stem()
+        .ok_or(anyhow!("the benchmark must have a file name"))?;
+    let name = if stem == "benchmark" {
+        path.parent()
+            .ok_or(anyhow!("the benchmark must have a parent directory"))?
+            .file_name()
+            .ok_or(anyhow!("the parent directory to have a name"))?
+    } else {
+        stem
+    };
+    Ok(to_string_lossy(name))
+}
+
+/// Simplify the benchmark path if possible; e.g.:
+/// `/home/user/code/sightglass/benchmarks-next/<name>/benchmark.wasm` ->
+/// `benchmarks-next/<name>/benchmark.wasm`. This function finds a path component matching
+/// `benchmarks/` or `benchmarks-next/` and cuts the path there.
+fn simplify_benchmark_path<P: AsRef<Path>>(path: P) -> String {
+    let path = path.as_ref();
+    if let Some(i) = path
+        .iter()
+        .position(|c| c == "benchmarks" || c == "benchmarks-next")
+    {
+        let shortened_path: PathBuf = path.iter().skip(i).collect();
+        to_string_lossy(shortened_path.as_os_str())
+    } else {
+        to_string_lossy(path.as_os_str())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn shortened_benchmark_names() {
+        assert_eq!(
+            simplify_benchmark_name("benchmarks-next/noop/benchmark.wasm").unwrap(),
+            "noop"
+        );
+        assert_eq!(simplify_benchmark_name("a/b/c.wasm").unwrap(), "c");
+    }
+
+    #[test]
+    fn shortened_benchmark_paths() {
+        assert_eq!(
+            simplify_benchmark_path("/home/user/sightglass/benchmarks/benchmark.wasm"),
+            format!("benchmarks{}benchmark.wasm", std::path::MAIN_SEPARATOR)
+        );
+        assert_eq!(
+            simplify_benchmark_path("code/benchmarks-next/noop.wasm"),
+            format!("benchmarks-next{}noop.wasm", std::path::MAIN_SEPARATOR)
+        );
+        // Note here how `simplify_benchmark_path` does not modify the path separator.
+        assert_eq!(
+            simplify_benchmark_path("some/other/path.wasm"),
+            "some/other/path.wasm"
+        );
```

Comment:
For the future: this is rather unexpected behavior, in that it isn't always clear when fingerprinting will use the OS path separator and when it will just pass it through. I'm going with "use the OS separator" instead of "replace everything to forward slashes" but I can see how in the future we might want to just use forward slashes to make comparing these things easier. Leaving this for later...